austation/code/modules/popes-xenoarchaeology/xenoartifact_console.dm

```diff
@@ -0,0 +1,271 @@
+            return
+
+    for(var/datum/xenoartifactseller/S in sellers)
+        if(action == "purchase_[S.unique_id]" && linked_inbox && budget.account_balance-S.price >= 0)
+            var/obj/item/xenoartifact/X = new(get_turf(linked_inbox.loc), S.difficulty)
+            X.price = S.price
+            sellers -= S
+            budget.adjust_money(-1*S.price)
+            say("Purchase complete. [budget.account_balance] credits remaining in Research Budget")
+            addtimer(CALLBACK(src, .proc/generate_new_seller), (rand(1,5)*60) SECONDS)
+            return
+        else if(action == "purchase_[S.unique_id]" && !linked_inbox)
+            say("Error. No linked hardware.")
+            return
+        else if(action == "purchase_[S.unique_id]" && budget.account_balance-S.price < 0)
+            say("Error. Insufficient funds.")
+            return
+
+    if(action == "sell")
+        if(!linked_inbox)
+            say("Error. No linked hardware.")
+            return
+        var/info
+        var/final_price = 100
+        var/obj/nearby = linked_inbox.get_item_prox(1)
+        for(var/obj/I in nearby)
+            var/avoidtimewaste = TRUE
+            for(var/datum/xenoartifactseller/buyer/B in buyers)//Check to avoid wasting time & to see if someone is actually buying that item.
+                if(istype(I, B.buying))
+                    avoidtimewaste = FALSE
+                    buyers -= B
+                    break
+            if(avoidtimewaste)
+                return
+            if(istype(I, /obj/item/xenoartifact)) //This and it's brother is pretty iffy
+                var/obj/item/xenoartifact/X = I
+                final_price = X.modifier*X.price
+                if(final_price < 0) //No modulate?
+                    final_price = X.price*0.1
+                budget.adjust_money(final_price)
+                linked_techweb.add_point_type(TECHWEB_POINT_TYPE_DEFAULT, final_price*10)
+                info = "[X.name] sold at [station_time_timestamp()] for [final_price] credits, bought for [X.price]"
+                sold_artifacts += list(info)
+                say(info)
+                qdel(I)
+                addtimer(CALLBACK(src, .proc/generate_new_buyer), (rand(1,5)*60) SECONDS)
+                return
+            else if(istype(I, /obj/structure/xenoartifact))
+                var/obj/structure/xenoartifact/X = I
+                final_price = X.modifier*X.price
+                if(final_price < 0)
+                    final_price = X.price*0.1
+                budget.adjust_money(final_price)
+                linked_techweb.add_point_type(TECHWEB_POINT_TYPE_DEFAULT, final_price*10)
+                info = "[X.name] sold at [station_time_timestamp()] for [final_price] credits, bought for [X.price]"
+                sold_artifacts += list(info)
+                say(info)
+                qdel(I)
+                addtimer(CALLBACK(src, .proc/generate_new_buyer), (rand(1,5)*60) SECONDS)
+                return
```

Comment:
You should make xeno artifacts a component so you don't need to specifically type the object and run duplicate code like this